### PR TITLE
Model setup

### PR DESCRIPTION
**Pull request by jhenley45**
Created models/controllers and set up basic associations. I left out the polymorphic, going to defer to you on that one.

I added two has many through relationships for users to establishments. My thinking is that I want to see both the establishments that a user has reviewed as well as the establishments they have visited, and these might not always be the same. Not sure if there is a built-in way to do this so I used the "source" option on two has many's. Basically you can call user.establishments_visited and user.establishments_reviewed. Is there a better way to do this?

### CODE REVIEW

File: rails/Gemfile [hunk 1]
```diff
@@ -1,40 +1,17 @@
...
-# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
 gem 'rails', '4.1.0'
-# Use postgresql as the database for Active Record
 gem 'pg'
-# Use SCSS for stylesheets
 gem 'sass-rails', '~> 4.0.3'
-# Use Uglifier as compressor for JavaScript assets
 gem 'uglifier', '>= 1.3.0'
-# Use CoffeeScript for .js.coffee assets and views
-gem 'coffee-rails', '~> 4.0.0'
```
Comment (mikestone14) [inline rails/Gemfile:13]: why'd you take out coffeescript? not feeling it?

File: rails/app/models/establishment.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class Establishment < ActiveRecord::Base
+	has_many :point_events
```
Comment (mikestone14) [inline rails/app/models/establishment.rb:2]: this indentation is friggin enormous bro.

File: rails/app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class User < ActiveRecord::Base
+	has_many :point_events
```
Comment (mikestone14) [inline rails/app/models/user.rb:2]: throw a few more indents in there while you're at it.

File: rails/Gemfile (hunk 1 above)
Comment (jhenley45) [inline rails/Gemfile:13]: Don't need it, all JS will be written and compiled in CLI.

Comment (jhenley45) [inline rails/Gemfile:13]: We can still do it in CLI if you want. I don't know it at all, could be something good to learn.

File: rails/test/controllers/reviews_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+require 'test_helper'
```
Comment (mikestone14) [inline rails/test/controllers/reviews_controller_test.rb:1]: holy moly, are we using test-unit? Not feeling rspec?

Comment (jhenley45) [inline rails/test/controllers/reviews_controller_test.rb:1]: I dunno bro I didn't touch this shit it's just the default. Rspec or die.

File: rails/app/models/point_event_type.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class PointEventType < ActiveRecord::Base
+	has_many :point_events
```
Comment (mikestone14) [inline rails/app/models/point_event_type.rb:2]: easy on the indent fella.

File: rails/app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+	has_many :visits
+	has_many :reviews
+	has_many :establishments_visited, through: :visits, :source => :establishment
```
Comment (mikestone14) [inline rails/app/models/user.rb:5]: I don't see why we need `establishments_visited` and `establishments_reviewed`. Waddupwidat?

File: rails/config/routes.rb [hunk 7]
```diff
@@ -1,4 +1,42 @@
 Rails.application.routes.draw do
+  get 'reviews/index'
```
Comment (mikestone14) [inline rails/config/routes.rb]: how about `resources :reviews`, etc.?

File: rails/app/models/point_event_type.rb (hunk 5 above)
Comment (jhenley45) [inline rails/app/models/point_event_type.rb:2]: One more of these and I'm gonna give you a knuckle sandwich.

File: rails/app/models/user.rb (hunk 6 above)
Comment (jhenley45) [inline rails/app/models/user.rb:5]: See my comment above for the commit. Wasn't entirely sure here.

File: rails/Gemfile (hunk 1 above)
Comment (mikestone14) [inline rails/Gemfile:13]: I think we should stick to jQuery in Ember. I guess if there are no views, etc. no reason for JS anyway.

File: rails/config/routes.rb (hunk 7 above)
Comment (jhenley45) [inline rails/config/routes.rb]: Ya, I didn't touch this yet.

File: rails/test/controllers/reviews_controller_test.rb (hunk 4 above)
Comment (mikestone14) [inline rails/test/controllers/reviews_controller_test.rb:1]: Ya dood eff this. What kind of generator are you using? I'm guessing it built all the controllers too. Whatever it is, it's whack.

File: rails/config/routes.rb (hunk 7 above)
Comment (mikestone14) [inline rails/config/routes.rb]: Not all these are get requests. Is this another generated thing?

File: rails/test/controllers/reviews_controller_test.rb (hunk 4 above)
Comment (jhenley45) [inline rails/test/controllers/reviews_controller_test.rb:1]: It's the default, I think it's just because I haven't added the rspec gem.

**Comment by mikestone14** [comment]
Re: `esblishments_visited` and `establishments_reviewed`, won't reviews belong to a user and an establishment? Similarly, visits will belong to a user and establishment. Since they will have both user and establishment identifiers as attributes I think we can get this info pretty easily.

**Comment by jhenley45** [comment]
How do you imagine getting at them? This seems pretty easy to me, unless there's a better/more direct way to do it.

**Comment by mikestone14** [comment]
Thanks for taking the initiative setting this up. Looks good. I think we can delete all that weird test stuff and set that stuff up in Rspec (in a separate PR). Also, are you familiar with rebasing? Basically it can be used to condense all of your commits for this feature into one feature commit to go into the PR. Makes it a bit easier I think. If you're not familiar try this:

1) `git checkout [branch_name]`
2) `git rebase -i origin/master`
3) This will pop you into an editor, either in the Terminal (Vim) or in Sublime. You will see all of your commits with the word "pick" in front of them. Keep the first "pick" and replace the rest of them with the word "squash". This will squash all of the other commits into the first one. You will have the option of editing the commit message later. Save and quit.
4) Once it's done, assuming there are no merge conflicts, it will again pop open an editor and you will see a list of your commit messages. Write an overall commit message describing the feature at the top, followed by whatever other information you want below. Save and quit.
5) `git push -f origin [branch_name]` The `-f` is to force push.
6) Keep on living your life.

**Comment by jhenley45** [comment]
Never tried it, here goes

**Comment by mikestone14** [comment]
To get establishments reviewed I guess we could do something like:

``` ruby
class Establishment

  def self.reviewed_by(user)
    user_reviews = Review.where(user: user)
    establishment_ids = user_reviews.select(&:establishment_id)
    Establishment.where(id: establishment_ids)
  end
end
```

How would you set it up otherwise? Create an `establishment_reviewed` record every time you create a `review` record?

**Comment by jhenley45** [comment]
This is what I was thinking and also what I'm trying to avoid. The line at the top accomplishes the same thing but with much less code.

The way I'm thinking about it is that there are two possible ways for a user to have a establishments; a user has many establishments through reviews and a user has many establishments through visits, and we should keep these relationships separate. This is just by nature of both users and establishments having many reviews and visits, and both reviews and visits belonging to both users and establishments (what? exactly). Imagine if it were just Users, Visits, and Establishments. You could easily say that a user has many establishments through visits, boom done... user.establishments. But in our case a user also has establishments through reviews, so we can't call user.establishments here as well. They are "through" different models, so I gave them different names.

**Comment by jhenley45** [comment]
Also currently stuck on 3.) in your explanation above. 

jackhenley@Jacks-MacBook-Pro:rails (model_setup)$ git rebase -i origin/master
fatal: Not a valid object name User
fatal: Needed a single revision
Invalid commit name: User
jackhenley@Jacks-MacBook-Pro:rails ((no branch, rebasing model_setup))$

What the?

**Comment by jhenley45** [comment]
Rebasing sucks

**Comment by mikestone14** [comment]
uhhhhh, :water_buffalo: not sure about the rebase thing, I've never seen that. I'll look into it and let you know. In the meantime I rebased for you and pushed it up.

**Comment by mikestone14** [comment]
Wait a sec, won't that mean that we have to have all establishments listed twice? You can have the same establishment as an establishment visited and an establishment reviewed? That doesn't seem to make sense.

**Comment by jhenley45** [comment]
Ya there's potentially some overlap but you should think of them as two separate collections with two separate purposes. One is to see all of the establishments that a user has visited, the other is to see all of the establishments that a user has reviewed.

**Comment by mikestone14** [comment]
It sounds like it's either 2 queries or 2 duplicate rows, right?

**Comment by mikestone14** [comment]
``` ruby
class User
  def establishments_reviewed
    Establishment.where(id: reviews.select(&:establishment_id))
  end
end
```

**Comment by mikestone14** [comment]
Did you clone or fork the repo?

**Comment by jhenley45** [comment]
OK, that's shorter, but whats the difference?

"It sounds like it's either 2 queries or 2 duplicate rows, right?" Not entirely sure what you mean here.

**Comment by mikestone14** [comment]
I don't like having duplicate rows for the same model basically.

There are users and there are establishments. The interaction between the 2 happens with a visit or a review. That doesn't change what an establishment is. To have separate models for reviewed establishments and visited establishments doesn't make sense to me. I think we can get the desired info very easily without it too.

**Comment by jhenley45** [comment]
I created the repo locally so neither

**Comment by mikestone14** [comment]
do you have to set the upstream or something?

**Comment by jhenley45** [comment]
We don't have any additional rows in the DB, and we're not creating any additional models. We're only using the pre-existing models- Users, establishments, reviews, and visits- and accessing establishments THROUGH reviews and visits (which will exist regardless of this "through" association). I think that's where you're getting tripped up.

**Comment by mikestone14** [comment]
Ok yeah that makes more sense. Maybe I'm just adverse to has_many through associations because I rarely see them and it seems like too much coupling. 

Are you able to have a has_many multiple times on the same model?

``` ruby
class User
  has_many :establishments, through: :visits
  has_many :establishments, through: :reviews
end
```

If that is the case, how would you get a collection of the establishments a user has reviewed versus visited?

It just seems a little funky and unnatural to me. Happy to do it if you want my gut is telling me no but that thing is never right.

**Comment by mikestone14** [comment]
Sorry, I see what you're doing using the `source`. Feels like too much coupling of the models but whatever you want. What happens if we add another way a user can interact with an establishment? Another has_many through?

**Comment by jhenley45** [comment]
Ya prolly. To me it does the same thing as the code that you posted just a bit neater. It's the same question- if we add another way a user can interact with an establishment do we add another method? Method vs relationship, doesn't make a big difference to me though.

**Comment by mikestone14** [comment]
Yeah I don't know enough to really know but methods are much easier to change around than database relationships. Once things are tied together it's really hard to untangle them. I think the less our models know about other models the better. Again, I don't know anything.

**Comment by jhenley45** [comment]
Haha ok. This is the way we set up the DB- reviews and visits both belong to establishments AND users. We can either handle the relationship via rails has_many_through or custom methods, but doing it w methods doesn't make it any less "coupled" in my opinion, it's just how our DB is. Either way, if we want to fundamentally change how users, reviews, visits, and establishments are related we are going to have to change the DB structure. 

Functionally there's 0 difference between a custom method and the has_many, they're both allowing us to easily access the relationships in our DB structure.

(we're off to a great start!)

**Comment by mikestone14** [comment]
Ok let's stick with has_many through. I'm really coming around to this idea. This might be helpful too (for me). http://stackoverflow.com/questions/11600928/when-should-one-use-a-has-many-through-relation-in-rails

**Comment by jhenley45** [comment]
Right, same thing except we have 2 of them.

Did you merge the model_setup branch and push?

Also we should create a development branch that will act as our working "master" branch. When we know development is stable we can tag a release and merge it into master, which will essentially be our production code. Keeps master clean.

**Comment by mikestone14** [comment]
Cool. I'm all about those has_many throughs. 

Once you clean up the rest of the garbage that those generators made I'll merge it. 

If the code is well tested we won't need a development branch. If all tests are green and our discussion is done, code looks good, then we merge to master and know master is clean.

**Comment by jhenley45** [comment]
OK, so new commit with changes will be a new push to model_setup and then a new pull request, no? Is there a way to add a commit to a current pull request?

**Comment by jhenley45** [comment]
Guess that answers that. They're just added to this pull request.

Not sure about rebasing, going to avoid it until I'm comfortable with it. Was doing some wacky shit.

**Comment by mikestone14** [comment]
Ok cool. Yeah if you make changes on his branch just commit them and they will be part of the PR. Also if you change any of the code on a line I commented on the comment will go away. Like changing the indentation on the models.

**Comment by jhenley45** [comment]
Bro dunno what to tell you that shit is commented normally in Sublime. Could just be our different editors... are you using Vim you old man you?

**Pull request merged**